File: apps/clojure/design-pattern/test/design_pattern/singleton_test.clj
(ns design-pattern.singleton-test
  (:require [clojure.test :refer :all]
            [design-pattern.singleton :refer :all]))

(deftest config-get-test
  (testing "設定値を取得できる"
    (is (= :info (get-config :log-level)))))

(deftest config-set-test
  (testing "設定値を更新できる"
    (let [original (get-config :log-level)]
      (set-config! :log-level :debug)
      (is (= :debug (get-config :log-level)))
      ;; 元に戻す
      (set-config! :log-level original))))

(deftest logger-test
  (testing "ログエントリを追加・取得できる"
    (clear-logs!)
    (log! :info "Test message")
    (log! :error "Error message")
    (is (= 2 (log-count)))
    (is (= :info (:level (first (get-logs)))))
    (clear-logs!)))

(deftest registry-test
  (testing "レジストリにサービスを登録・取得できる"
    (clear-registry!)
    (register! :db {:host "localhost" :port 5432})
    (is (= {:host "localhost" :port 5432} (lookup :db)))
    (clear-registry!)))

(deftest singleton-identity-test
  (testing "同じ atom への参照を共有している（シングルトン性）"
    (clear-logs!)
    (log! :info "first")
    (is (= 1 (log-count)))
    (log! :info "second")
    (is (= 2 (log-count)))
    (clear-logs!)))

File: apps/elixir/design-pattern/test/design_pattern/singleton_test.exs
defmodule DesignPattern.SingletonTest do
  use ExUnit.Case
  alias DesignPattern.Singleton
  alias DesignPattern.Singleton.Registry

  test "設定値を保存・取得できる" do
    Singleton.put(:db_host, "localhost")
    assert Singleton.get(:db_host) == "localhost"
    Singleton.delete(:db_host)
  end

  test "存在しないキーにはデフォルト値を返す" do
    assert Singleton.get(:nonexistent, "default") == "default"
  end

  test "設定値を削除できる" do
    Singleton.put(:temp_key, "value")
    Singleton.delete(:temp_key)
    assert Singleton.get(:temp_key) == nil
  end

  test "設定値を上書きできる" do
    Singleton.put(:counter, 1)
    Singleton.put(:counter, 2)
    assert Singleton.get(:counter) == 2
    Singleton.delete(:counter)
  end

  test "Registry に値を登録・参照できる" do
    Registry.register(:logger, %{level: :info})
    assert {:ok, %{level: :info}} = Registry.lookup(:logger)
    Registry.unregister(:logger)
  end

  test "Registry で未登録のキーは error を返す" do
    assert {:error, :not_found} = Registry.lookup(:unknown)
  end
end
